cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -268,79 +284,431 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+// - failed to find data at all -> try to unlock save
+// - unlocked data with run key -> mount data
+func (m *stateMachine) unlockDataRunKey(ctx *recoverContext) (stateFunc, error) {
 	runModeKey := filepath.Join(boot.InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key")
-	device, isDecryptDev, err := secbootUnlockVolumeUsingSealedKeyIfEncrypted(disk, "ubuntu-data", runModeKey, lockKeysOnFinish)
+	dataDevice, isEncryptedDev, err := secbootUnlockVolumeUsingSealedKeyIfEncrypted(ctx.disk, "ubuntu-data", runModeKey, false)
+	if isEncryptedDev {
+		ctx.diskOpts = &disks.Options{
+			IsDecryptedDevice: true,
+		}
+		ctx.isEncryptedDev = true
+	}
 	if err != nil {
-		return err
+		// we couldn't unlock ubuntu-data with the primary key, or we didn't
+		// find it in the unencrypted case
+		if isEncryptedDev {
+			ctx.degradedState.LogErrorf("cannot unlock encrypted ubuntu-data with sealed run key: %v", err)
+			// we know the device is encrypted, so the next state is to try
+			// unlocking with the fallback key
+			return m.unlockDataFallbackKey, nil
+		}
+
+		// not an encrypted device, so nothing to fall back to try and unlock
+		// data, so just mark it as not found and continue on to try and mount
+		// an unencrypted ubuntu-save directly
+		// TODO:UC20: should we save the specific error in degradedState
+		//            somewhere in addition to logging it?
+		ctx.degradedState.LogErrorf("failed to find ubuntu-data partition for mounting host data: %v", err)
+		ctx.degradedState.DataState = "other"
+		return m.locateUnencryptedSave, nil
 	}
 
+	// otherwise successfully unlocked it (if it was encrypted)
+	ctx.dataDevice = dataDevice
+
+	// successfully unlocked it with the run key, so just mark that in the
+	// state and move on to trying to mount it
+	if isEncryptedDev {
+		ctx.degradedState.DataKey = "run"
+	}
+
+	return m.mountData, nil
+}
+
+func (m *stateMachine) unlockDataFallbackKey(ctx *recoverContext) (stateFunc, error) {
+	dataFallbackKey := filepath.Join(boot.InitramfsEncryptionKeyDir, "ubuntu-data.recovery.sealed-key")
+	// XXX: we don't check isDecryptDev here, if we are here then the previous
+	// call trying to unlock ubuntu-data already said it was encrypted
+	dataDevice, _, err := secbootUnlockVolumeUsingSealedKeyIfEncrypted(ctx.disk, "ubuntu-data", dataFallbackKey, false)
+	if err != nil {
+		// TODO: should we introspect err here to get a more detailed
+		//       response in degradedState and in the log message?
+		// we failed to decrypt the device again with the fallback key,
+		// so we are for sure in degraded mode
+		ctx.degradedState.LogErrorf("cannot find or unlock encrypted ubuntu-data partition with sealed fallback key: %v", err)
+		ctx.degradedState.DataState = "enc-not-found"
+
+		// skip trying to mount data, since we did not data we cannot
+		// open save with with the run key, so try the fallback one
+		return m.unlockSaveFallbackKey, nil
```

Comment:
at this point try the data recovery key?

Reply:
this is done then